exchange.go

```diff
@@ -442,7 +475,144 @@ func (re *RabbitExchangeImpl) Receive(exchange ExchangeSettings, queue QueueSett
 						closer()
 						for {
 							time.Sleep(time.Millisecond * 250)
-							channel, msgs, errChan, closer, err = getChannel()
+							channel, msgs, errChan, closer, err = re.getChannel(exchange, queue, retryExchangeName)
+							if err != nil {
+								log.Printf("Error connecting to rabbit %+v\n", err)
+							} else {
+								break
+							}
+						}
+					}
+
+				}
+			}
+		},
+		func() {
+			close(stop)
+		}, nil
+}
+
+func (re *RabbitExchangeImpl) BulkReceive(exchange ExchangeSettings, queue QueueSettings, maxWait int) (func(BulkMessageHandleFunc) error, func(), error) {
+	retryExchangeName := fmt.Sprintf("%s-%s", exchange.Name, retryExchangeNameSuffix)
+
+	channel, msgs, errChan, closer, err := re.getChannel(exchange, queue, retryExchangeName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	stop := make(chan struct{})
+	return func(handler BulkMessageHandleFunc) error {
+			defer closer()
+			batch := make([]amqp.Delivery, 0, queue.Prefetch)
+			timerDuration := time.Duration(maxWait) * time.Millisecond
+			fillWaitTimer := time.NewTimer(timerDuration)
+			handleMessages := func(messages []amqp.Delivery) {
+				if len(messages) == 0 {
+					return
+				}
+				ctx, cancel := context.WithCancel(context.Background())
+				defer cancel()
+				errors := handler(ctx, messages)
+				erroredMessages := make(map[string]struct{})
+				if len(errors) > 0 {
+					for _, msgError := range errors {
+						m := msgError.message
+						err := msgError.err
+						erroredMessages[m.MessageId] = struct{}{}
+						log.Printf("Error handling rabbit message Exchange: %s Queue: %s Body: [%s] %+v\n", exchange.Name, queue.Name, m.Body, err)
+						isProcessingError := IsTemporaryError(err)
+						err = m.Nack(false, false)
+						if err != nil {
+							log.Printf("Error Nack rabbit message %+v\n", err)
+						}
+						if isProcessingError {
+							requeuesObj, ok := m.Headers[requeueHeaderKey]
+							if !ok {
+								continue
+							}
+							noRequeues, ok := requeuesObj.(int32)
+							if ok && noRequeues < maxRequeueNo {
+								noRequeues++
+								err = requeueMessage(
+									channel,
+									queue,
+									exchange.Name,
+									retryExchangeName,
+									m.Body,
+									int(noRequeues),
+								)
+								if err != nil {
+									log.Printf("Error requeueing message %+v\n", err)
+								}
+							}
+						}
+					}
+					for _, m := range messages {
+						if _, ok := erroredMessages[m.MessageId]; !ok {
+							m.Ack(false)
+						}
+					}
+				} else {
+					// If none of the messages resulted in errors, ack multiple (the whole prefetched batch)
+					err = messages[0].Ack(true)
```

Comment:
Is this the right message? 
Message 0 was the first message we received, so it is the oldest. 
I think Ack(true) acks all message older than the message you ack, so this will only Ack 1 message (the oldest one).

We want to Ack the newest one, so it can mark all the others as handled (but I will go read docs). 

Either way, we should convince the reader with a comment :)

Reply:
> When multiple is true, this delivery and all prior unacknowledged deliveries
on the same channel will be acknowledged.  This is useful for batch processing
of deliveries.

Yeah, so we need to Ack the last message in the batch so it can ack all prior ones.

Reply:
Oh boy, thanks for the catch.